org.eclipse.xtext/src/org/eclipse/xtext/formatting2/AbstractJavaFormatter.java

```diff
@@ -0,0 +1,182 @@
+ * 		for (Parent parent : model.getParents())
+ * 			doc.format(parent);
+ * 	}
+ * 
+ * 	protected void _format(Parent parent, IFormattableDocument doc) {
+ * 		doc.prepend(regionFor(parent).keyword("parent"), it -> it.noSpace());
+ * 		doc.append(regionFor(parent).keyword("parent"), it -> it.oneSpace());
+ * 		doc.append(regionFor(parent).feature(PARENT__NAME), it -> it.oneSpace());
+ * 		doc.prepend(regionFor(parent).keyword("{"), it -> it.oneSpace());
+ * 		doc.append(regionFor(parent).keyword("{"), it -> it.newLine());
+ * 		doc.interior(regionFor(parent).keyword("{"), regionFor(parent).keyword("}"), it -> it.indent());
+ * 		doc.append(regionFor(parent).keyword("}"), it -> it.setNewLines(1, 1, 2));
+ * 		for (Child child : parent.getChildren())
+ * 			doc.format(child);
+ * 	}
+ * 
+ * 	protected void _format(Child child, IFormattableDocument doc) {
+ * 		doc.append(regionFor(child).keyword("child"), it -> it.oneSpace());
+ * 		doc.append(regionFor(child).feature(CHILD__NAME), it -> it.newLine());
+ * 	}
+ * }
+ * </pre>
+ * </p>
+ * <p>
+ * This is suitable for the following grammar:
+ * 
+ * <pre>
+ * grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
+ * 
+ * generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
+ * 
+ * Model: parents+=Parent*;
+ * Parent: 'parent' name=ID '!' ('{' children+=Child* '}')?;
+ * Child: 'child' name=ID;
+ * </pre>
+ * </p>
+ */
+public abstract class AbstractJavaFormatter extends AbstractFormatter2 implements ITextRegionExtensions {
+
+	// reflective method that calls "_format" methods found in the implementing class.
+	@Override
+	public void format(Object child, IFormattableDocument document) {
+		if (child instanceof XtextResource) {
+			_format((XtextResource) child, document);
+			return;
+		} else if (child == null) {
+			_format((Void) null, document);
+			return;
+		}
+		try {
+			Method[] methods = getClass().getDeclaredMethods();
+			for (Method _format : methods) {
+				if (!_format.getName().equals("_format"))
+					continue;
+				if (_format.getParameterCount() != 2)
+					continue;
+				if (_format.getParameters()[1].getType() != IFormattableDocument.class)
+					continue;
+				if (!_format.getParameters()[0].getType().isInstance(child))
+					continue;
+				_format.setAccessible(true);
+				_format.invoke(this, child, document);
+				return;
+			}
+			if (child instanceof EObject) {
+				_format((EObject) child, document);
+				return;
+			} else {
+				_format(child, document);
+				return;
+			}
+		} catch (SecurityException | IllegalAccessException | IllegalArgumentException e) {
+			throw new RuntimeException(e);
+		} catch (InvocationTargetException e) {
+			Throwable cause = e.getCause();
+			if (cause instanceof RuntimeException)
+				throw (RuntimeException) cause;
+			throw new RuntimeException(cause);
+		}
+	}
+
+	// implementations that forward the methods of ITextRegionExtensions to simplify formatter code.
```

Comment:
should these be protected?

Reply:
Not possible. These are implementations from an interface.

Reply:
Do we strictly need to implement the interface? Can we implement the methods here as protected and iff we need an instance of ITextRegionAccess, we could use a method `textRegionAccess() { returns .. }` that creates an anonymous instance

Reply:
Guess that sould be possible ... will try.